### PR TITLE
Prettify EPUB internal file structure and contents

### PR DESCRIPTION
**Pull request by twilight-flower**
Changes to EPUB internal structure (if these impact reader display of the books at all, something has gone wrong):
- Change the naming formula for chapter/section XHTML files from `chapter{chapter}_{section}.html` to `{chapter}-{section} ({title}).xhtml`, and add padding to chapter and section numbers
    - Add validation to filter out title characters which are disallowed in EPUB 3 and to truncate filenames which, as a result of long thread titles, end up over 255 bytes in length post-filtering
- Change the naming formula for icon image files from `img{number}.{ext}` to `icon{number}.{ext}`, and add zero-padding to image the numbers.
- Store all chapter/section XHTML files in `EPUB/Text/` instead of in `EPUB/`
- Store all image files in `EPUB/Images/` instead of in `EPUB/`

Low-level changes to individual files' code:
- Add newline to end of `style.css`
- Add alt text to all icon images, in accordance with XHTML requirements
- Add non-empty id to the top-level post of each thread (which has no permalink fragment), in accordance with XHTML requirements
- Run XHTML through the `lxml` pretty-printer in order to make the files more readable

Commentary:
- The extension change on the XHTML files is for compliance with the recommendations of the EPUB specification; the rest of the change to their naming scheme is to increase clarity for anyone manually browsing the EPUB's internal file tree.
- The name change from `img` to `icon` on the icon files is in preparation for a future PR, where I plan to add non-icon inline images into the downloads, rather than having those remain as external links. The zero-padding is for readability.
- In an ideal world, we'd also add useful semantic information into the icon filenames as we have to the XHTML filenames, more along the lines of e.g. `icon{number}_{gallery}_{title}.{ext}`. But I don't think there's any way to do so without adding in extra scraping of gallery pages, and the costs of that are likely not worth the reward.
- The XHTML pretty-printing is possibly going to make it start throwing errors when fed ill-formed XML? The issue hasn't arisen in any of several test runs, though, so I suspect that some combination of the limits of the Constellation editor and the repairs performed by the BeautifulSoup parser are sufficient to ensure that all XHTML fed into the pretty-printer is well-formed.
    - (When I tried to feed ill-formed HTML into the Constellation's editor, it automatically closed the tag I'd left open; perhaps someone cleverer could bypass whatever protections are in place, but that was sufficient testing to satisfy me, when coupled with the problem-free test runs.)
- All of these changes push in the general direction of making the EPUBs marginally larger. Longer filenames, more whitespace in the XHTML files. But the change is negligible compared with the sizes of e.g. icon image files, such that I'm not nearly as worried about the extra size as I would be under other circumstances. (If size is a major constraint, even the files in their current form will likely be too much, and if it's a weaker constraint, the marginal added size is unlikely to be large enough to make the difference.)

### CODE REVIEW

File: glowfic-dl.py [hunk 1]
```diff
@@ -130,11 +161,75 @@ def __init__(self, stamped_urls: list[StampedURL], title: str):
...
+        ext = split_filename[-1]
+        ext_bytes = len(ext.encode("utf-8"))
+        if ext_bytes > 255:
+            raise Exception(
+                "Attempted to put file into EPUB with extension longer than 255 bytes."
+            )
+
+        name_truncated = ".".join(split_filename[:-1])[:-1]
+        full_name_truncated = "%s.%s" % (name_truncated, ext)
+        while len(full_name_truncated.encode("utf-8")) > 255:
```
Comment (rocurley) [inline glowfic-dl.py]: To make sure I'm understanding what's going on here: you're lopping unicode characters off the end one at a time until the length in bytes is less than 255. You're doing this because it's unclear how many characters need to be removed from the end to remove a specific number of bytes. Is that right?

File: glowfic-dl.py [hunk 2]
```diff
@@ -158,13 +253,18 @@ def render_post(post: Tag, image_map: ImageMap) -> RenderedPost:
     post_div = post_html.find("div")
     permalink = post.find("img", title="Permalink", alt="Permalink").parent["href"]
     permalink_fragment = urlparse(permalink).fragment
-    reply_anchor = post_html.new_tag("a", id=permalink_fragment)
+    if permalink_fragment == "":
```
Comment (rocurley) [inline glowfic-dl.py]: I don't think the empty-fragment anchors were actually used anywhere, so if you're going to check for them, is there any reason not to just drop them entirely?

File: glowfic-dl.py [hunk 3]
```diff
@@ -207,13 +308,30 @@ async def download_chapter(
     return (title, list(render_posts(posts, image_map, authors)))
 
 
-def compile_chapters(chapters: list[tuple[str, list[Section]]]) -> list[epub.EpubHtml]:
+def compile_chapters(
+    chapters: list[tuple[str, list[Section]]]
+) -> Iterable[list[epub.EpubHtml]]:
+    chapter_digits = len(str(len(chapters)))
```
Comment (rocurley) [inline glowfic-dl.py]: Aren't we zero-indexing? Which is to say, with 10 chapters, the greatest id would be 9, not 10.

File: glowfic-dl.py [hunk 4]
```diff
@@ -207,13 +308,30 @@ async def download_chapter(
...
     anchor_sections = {}
+
+    # Map permalinks to file names
     for (i, (title, sections)) in enumerate(chapters):
+        section_digits = len(str(len(sections)))
         for (j, section) in enumerate(sections):
-            file_name = "chapter%i_%i.html" % (i, j)
+            file_name = "Text/" + make_filename_valid_for_epub3(
```
Comment (rocurley) [inline glowfic-dl.py]: Doesn't "Text/" cut into the 255 character limit?

File: glowfic-dl.py [hunk 5]
```diff
@@ -207,13 +308,30 @@ async def download_chapter(
...
+                    i + 1,
+                    section_digits,
+                    j,
+                    title,
+                )
+            )
             for permalink in section.link_targets:
                 anchor_sections[permalink] = file_name
+
+    # Replace external links with internal links where practical
```
Comment (rocurley) [inline glowfic-dl.py]: Nit: Maybe "possible"? "Practicial" to me implies that there are some links we could convert in theory but don't because it's too hard.

File: glowfic-dl.py [hunk 6]
```diff
@@ -227,16 +345,34 @@ def compile_chapters(chapters: list[tuple[str, list[Section]]]) -> list[epub.Epu
                     a["href"] = url._replace(
                         scheme="https", netloc="glowfic.com"
                     ).geturl()
+
+    # Yield one list of EpubHTML objects per chapter
     for (i, (title, sections)) in enumerate(chapters):
+        section_digits = len(str(len(sections)))
```
Comment (rocurley) [inline glowfic-dl.py]: See `chatper_digits` comment.

File: glowfic-dl.py [hunk 7]
```diff
@@ -339,7 +477,7 @@ def get_cookies() -> dict[str, str]:
             (name, cookie) = raw.split("=")
             if name != COOKIE_NAME:
                 raise ValueError(
-                    f'cookie file must start with "{COOKIE_NAME}=" (no quotes)'
+                    'cookie file must start with "%s=" (no quotes)' % COOKIE_NAME
```
Comment (rocurley) [inline glowfic-dl.py]: No objection per se but why this change?

File: glowfic-dl.py [hunk 8]
```diff
@@ -130,11 +161,75 @@ def __init__(self, stamped_urls: list[StampedURL], title: str):
         self.last_update = max((stamped_url.stamp for stamped_url in stamped_urls))
 
 
+#################
+##   Helpers   ##
+#################
+
+
+def make_filename_valid_for_epub3(filename: str) -> str:
```
Comment (rocurley) [inline glowfic-dl.py]: Two concerns here:

* Is this something ebooklib handles? Probably worth checking if you haven't.
* Could you write a few unit tests for this? I don't want to be too much of a stickler on a small project like this but this function is pretty complicated and also rarely exercised.

**Review by rocurley — COMMENTED** [review]
A few small things but LGTM overall.

File: glowfic-dl.py (hunk 4 above)
Comment (twilight-flower) [reply glowfic-dl.py]: It doesn't. The 255-byte limit is on each path component, not on the full path. (There's a 65535-byte limit on total path length which "Text/" _does_ cut into, but there's currently no risk of colliding with that one, since everything but the final filename is hardcoded.)

File: glowfic-dl.py (hunk 1 above)
Comment (twilight-flower) [reply glowfic-dl.py]: Your understanding is correct, yup!

File: glowfic-dl.py (hunk 7 above)
Comment (twilight-flower) [reply glowfic-dl.py]: Stylistic consistency. I used the f-string in my last PR out of habit, when refactoring `get_cookies` out into its own function, but that made it the only string in the whole script to use an f-string rather than the mod operator, and the inconsistency was bugging me, so I changed it to be in line with the rest of the script.

File: glowfic-dl.py (hunk 3 above)
Comment (twilight-flower) [reply glowfic-dl.py]: I switched the chapter numbers to be 1-indexed as part of the general "increase filename human-readability" effort. I can switch them back if you'd rather stick with consistent zero-indexing, though. (And you're right that I made a mistake in this regard in the section_digits case, since that one _is_ meant to still be 0-indexed.)

File: glowfic-dl.py (hunk 2 above)
Comment (twilight-flower) [reply glowfic-dl.py]: None that I can think of! I have now done so.

File: glowfic-dl.py (hunk 5 above)
Comment (twilight-flower) [reply glowfic-dl.py]: I've updated the phrasing. And also discovered it's not actually accurate, currently; apparently some link-replacement actually fails right now. (See the [planecrash lectures](https://glowfic.com/board_sections/721) section, for a lightweight example: the link at the end of `to hell with science`, in the downloaded EPUB, goes out to glowfic.com rather than going to the book's internal `to earth with science` copy.) So I'll add that to my pile of things to try to fix in my next PR.

File: glowfic-dl.py (hunk 6 above)
Comment (twilight-flower) [reply glowfic-dl.py]: Fixed in the latest commit.

File: glowfic-dl.py (hunk 8 above)
Comment (twilight-flower) [reply glowfic-dl.py]: ebooklib seems not to quite do _nothing_ in this field, but neither does what it _does_ do produce well-formed filenames. It seems to, for whatever reason, truncate filenames down to 259 characters (or possibly bytes, I only tested with single-byte characters) instead of 255? And it does nothing to filter out disallowed characters such as DEL. So the extra filtering step definitely does seem to be needed. I'll get started on some basic tests.

Comment (twilight-flower) [reply glowfic-dl.py]: Basic unit testing is now in place! Setting it up required refactoring the code from one module into several, in order to avoid difficulties importing `glowfic-dl.py` into the test file in light of the hyphen in its name, so that's now done too.

(Workarounds in theory could have been used to import it even given the hyphen's presence—`importlib.import_module`, for instance—but I suspect they'd have ended up as technical debt in the long run, whereas splitting into modules is probably healthy for the long-term shape of the project _anyway_ as it expands.)

**Review by rocurley — APPROVED** [review]
Lgtm. Not merging yet in case there was anything else you were in the middle of. Let me know and I'll merge it

**Comment by twilight-flower** [comment]
Merging now sounds good to me!
